### PR TITLE
metrics: make gauge_float64 and counter_float64 lock free

### PR DESCRIPTION
**Pull request by Exca-DK**
This PR changes the gauge and counter of type f64 to use atomic instead of mutex.

BenchmarkCounterFloat64Pararel-32       186.0 ns/op
BenchmarkCounterFloat64Pararel-32       169.3 ns/op

BenchmarkGaugeFloat64Pararel-32         180.7 ns/op
BenchmarkGaugeFloat64Pararel-32         170.2 ns/op

### CODE REVIEW

File: metrics/counter_float64.go [hunk 1]
```diff
@@ -115,39 +116,52 @@ func (NilCounterFloat64) Snapshot() CounterFloat64 { return NilCounterFloat64{}
 // StandardCounterFloat64 is the standard implementation of a CounterFloat64 and uses the
 // sync.Mutex package to manage a single float64 value.
 type StandardCounterFloat64 struct {
-	mutex sync.Mutex
-	count float64
+	floatBits uint64
```
Comment (holiman) [inline metrics/counter_float64.go:119]: Why not use the new `atomic.Uint64` instead?

File: metrics/counter_float_64_test.go [hunk 2]
```diff
@@ -10,6 +13,20 @@ func BenchmarkCounterFloat64(b *testing.B) {
 	}
 }
 
+func BenchmarkCounterFloat64Pararel(b *testing.B) {
```
Comment (holiman) [inline metrics/counter_float_64_test.go:16]:
```suggestion
func BenchmarkCounterFloat64Parallel(b *testing.B) {
```

**Review by holiman — COMMENTED** [review]
Elegant idea! Lgtm, except some nitpicks. 
Also, it would be interesting to see a before/after benchmark.

**Comment by holiman** [comment]
```
[user@work metrics]$ benchstat float.1 float.2
name              old time/op  new time/op  delta
CounterFloat64-8  36.1ns ± 7%  21.2ns ±11%  -41.45%  (p=0.008 n=5+5)
GaugeFloat64-8    32.3ns ±27%  14.3ns ±23%  -55.63%  (p=0.008 n=5+5)
```
:+1:

File: metrics/counter_float64.go (hunk 1 above)
Comment (Exca-DK) [reply metrics/counter_float64.go:119]: Because it would result in one extra & operation to pass it to helper functions.

Comment (holiman) [reply metrics/counter_float64.go:119]: For the sake of clarity and adhering to what nowadays is the go-lang recommended practice, please switch to use `atomic.Uint64` instead.

Comment (holiman) [reply metrics/counter_float64.go:119]: Unless I am missing something the overhead can't be that bad (?)

File: metrics/counter_float_64_test.go [hunk 3]
```diff
@@ -10,6 +13,20 @@ func BenchmarkCounterFloat64(b *testing.B) {
...
+	var wg sync.WaitGroup
+	for i := 0; i < b.N; i++ {
+		wg.Add(1)
+		go func(i float64) {
+			c.Inc(i)
+			wg.Done()
+		}(float64(i))
+	}
+	wg.Wait()
+}
```
Comment (holiman) [inline metrics/counter_float_64_test.go:28]: I think this benchmark is a bit too parallel -- I suspect there's quite some noise from just maintaining `b.N` goroutines.

Perhaps a more realistic benchmark wold be do spin up 10 routines, and all ten routines does `b.N` calls to `c.Inc`.

**Comment by holiman** [comment]
I pushed some changes now, removing some very small utility methods which I just inlined instead. The way you had written the benchmark, using runParallel doesn't seem to work quite as one might expect. When I tested that before vs after, it showed a degradation. 
I made a more straight-forward benchmark instead, which just mimics ten writers trying to update the counter. That bench shows improvement as expected

```
name                      old time/op  new time/op  delta
CounterFloat64Parallel-8  1.45µs ±10%  0.85µs ± 6%  -41.65%  (p=0.008 n=5+5)

```

**Review by holiman — APPROVED** [review]
LGTM

**Pull request merged**